src/wazuh_db/wdb_global.c

```diff
@@ -1415,6 +1415,251 @@ cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id
+
+    if(!dp) {
+        mdebug1("Unable to create backup directory '%s' for Wazuh-DB", WDB_BACKUP_FOLDER);
+        snprintf(output, OS_MAXSTR + 1, "Unable to create backup directory '%s' for Wazuh-DB", WDB_BACKUP_FOLDER);
+        return OS_INVALID;
+    }
+    closedir(dp);
+
+    timestamp = w_get_timestamp(time(NULL));
+    wchr_replace(timestamp, ' ', '-');
+    wchr_replace(timestamp, '/', '-');
+    snprintf(path, PATH_MAX-3, "%s/%s.db-backup-%s", WDB_BACKUP_FOLDER, WDB_GLOB_NAME, timestamp);
+    os_free(timestamp);
+
+    // We can't run a VACUUM withing a transaction
+    if (wdb->transaction) {
+        if(wdb_commit2(wdb) == OS_INVALID) {
+            mdebug1("Cannot commit current transaction to create backup");
+            snprintf(output, OS_MAXSTR + 1, "Cannot commit current transaction to create backup");
+            return OS_INVALID;
+        }
+    }
+
+    if (wdb_stmt_cache(wdb, WDB_STMT_GLOBAL_VACUUM_INTO) < 0) {
+        mdebug1("Cannot cache statement");
+        snprintf(output, OS_MAXSTR + 1, "Cannot cache statement");
+        return OS_INVALID;
+    }
+
+    stmt = wdb->stmt[WDB_STMT_GLOBAL_VACUUM_INTO];
+
+    if (sqlite3_bind_text(stmt, 1, path , -1, NULL) != SQLITE_OK) {
+        merror("DB(%s) sqlite3_bind_text(): %s", wdb->id, sqlite3_errmsg(wdb->db));
+        snprintf(output, OS_MAXSTR + 1, "DB(%s) sqlite3_bind_text(): %s", wdb->id, sqlite3_errmsg(wdb->db));
+        return OS_INVALID;
+    }
+
+    switch (wdb_step(stmt)) {
+    case SQLITE_ROW:
+    case SQLITE_DONE:
+        result = OS_SUCCESS;
+        break;
+    default:
+        mdebug1("SQLite: %s", sqlite3_errmsg(wdb->db));
+        snprintf(output, OS_MAXSTR + 1, "SQLite: %s", sqlite3_errmsg(wdb->db));
+        result = OS_INVALID;
+    }
+
+    if(OS_SUCCESS == result) {
+        snprintf(path_compressed, PATH_MAX, "%s.gz", path);
+        result = w_compress_gzfile(path, path_compressed);
+        if(OS_SUCCESS == result) {
+            unlink(path);
+            wdb_remove_old_backup();
+        } else {
+            snprintf(output, OS_MAXSTR + 1, "Failed during database backup compression");
+        }
+    }
+
+    return result;
+}
+
+int wdb_remove_old_backup() {
+    DIR* dp = NULL;
+    struct dirent *entry = NULL;
+    time_t current_time = 0;
+    struct stat backup_info = {0};
+    int number_of_files = 0;
+    char tmp_path[OS_SIZE_512] = {0};
+    char** backups_array = NULL;
+    time_t* backups_time_array = NULL;
+    int backups_to_delete = 0;
+
+    dp = opendir(WDB_BACKUP_FOLDER);
+
+    if(!dp) {
+        mdebug1("Unable to open backup directory '%s'", WDB_BACKUP_FOLDER);
+        return OS_INVALID;
+    }
+
+    current_time = time(NULL);
+
+    // The pre-restore snapshot is not considered for the max_files limit
+    while (entry = readdir(dp), entry) {
+        if ((strcmp(entry->d_name, ".") == 0) || (strcmp(entry->d_name, "..") == 0) ||  (strcmp(entry->d_name, "global.db-backup-pre_restore.gz") == 0)) {
```

Comment:
It would be easier if the backup files has a unique path format, that could be matched with strncmp(). Doing so, we can avoid comparing for the pre_restore file, and any possible file found in the folder will not affect this logic.

Reply:
The `strncmp()` trick was implemented!